### ISSUE TITLE
Subscribing vs Callbacks from the DFSP

### ISSUE THREAD

**Issue by lewisdaly** [mojaloop/pisp#36]
> Ok. Sorry about the length of this one - I wanted to make sure I had clearly outlined the issues I see with our current thinking, so it's certainly not a brief read.

### Background:
This issue is a continuation on some of our conversations on slack and in calls around the method for informing a PISP of the transaction outcome.

In our current thinking, the _switch_ is responsible for sending a `PUT thirdpartyRequests/transactions/{id}` to the _PISP_ once it sees the `PUT /transfers/{id}` from the payee DFSP with a `transferState` of `COMMITTED` or `RESERVED`.

The method for doing such a callback is currently undefined, but the thinking so far has been that the switch can _subscribe_ to certain _events_ based on the requests that pass through it.

For delivering the `PUT thirdpartyRequests/transactions/{id}`, the switch would subscribe to _events_ related to a Transfer Request (specifically the `PUT /transfers/{id}` callback from the _Payee DFSP_). 
> Sidebar: We aren't 100% sure on the format or requirements for the subscription, such as how the switch will know that a certain `PUT /transfers/{id}` request related to a  `transactionRequestId`, but that's ok for now.

### Concerns About This Approach
The more thinking that I have been doing about error conditions within the context of a `thirdpartyRequests/transactions` between a PISP and DFSP, the less convinced I am that the _switch_ should be sending the `PUT thirdpartyRequests/transactions/{id}` to the PISP.

If we rely on the switch to be responsible for the `PUT thirdpartyRequests/transactions/{id}` to inform the PISP of a successful 3rd Party Transaction, it stands to reason that we should rely on the switch for informing the PISP of the failure conditions. For now let's assume that we are going to use the a `PUT thirdpartyRequests/transactions/{id}/error` request to do such informing.

This means that the switch would need to know any number of ways in which the 3rd Party Transaction could fail. Here's a non-exhaustive list off the top of my head:

- 1.1 The `Consent` provided by the PISP is not active (or maybe the credential has been revoked)
- 1.2 The `consentId` provided by the PISP is invalid or incorrect
- 1.3 PayeeDFSP's user rejects the quote
- 1.4 The signed `QuoteResponse` from the PISP was invalid, or signed with the wrong private key
- 1.5 The user has not enough funds in their account (specified by `sourceAccountId`) to complete the transaction
- 1.6 The transfer between the PayerDFSP and PayeeDFSP fails (maybe one participants position exceeded their net debit cap, or the transfer times out waiting for the PayeeDFSP to respond)

Let's ignore `1.4` above, as that could be handled by the `PUT /authorizations/{id}/error` callback, which the PayerDFSP will issue.

To evaluate the conditions `1.1`, `1.2`, and `1.5` _only_ the PayerDFSP can evaluate, which implies that the PayerDFSP must be responsible for sending a `PUT thirdpartyRequests/transactions/{id}/error` to the PISP.

`1.3` and `1.6` are where it gets interesting though.  Now (I hear your voice in the back of my head @MichaelJBRichards), one approach here is to also handle such callbacks with the _subscriptions_ concept as above, and have the PISP subscribe to events such as:
```
- PUT /quotes/{id}
- PUT /quotes/{id}/error
- PUT /transfers/{id} #this was already subscribed as above
- PUT /transfers/{id}/error
```
> **Another Sidebar:** This approach could actually do away with the need for a PayerDFSP to issue a`POST /authorizations`, since the PISP would be informed about the quote already... But let's not talk about that just yet.

My concern here is the _implicit_ behaviour of such a subscription model, and where the business logic for such behaviour would live.

The two options are:
1. The PISP receives the above callbacks, and interprets them (e.g. if it `PUT /quotes/{id}/error`, it implies that the `thirdpartyRequests/transactions` it issued earlier failed).
2. The Switch interprets the above events on `Quotes` and `Transfers` itself and send a `PUT thirdpartyRequests/transactions/{id}/error`. This means the Switch needs to hold a certain amount of business logic to do such an interpretation, and still not cover 100% of cases where a 3rd Party Request failed.

Additionally, from the PISP's perspective, there is no real source of truth for the state of a `thirdpartyRequests/transactions` - that source of truth is based on it's own "cobbling together" of a variety of callbacks from events emitted from the switch (be they derived from option 1 or 2 above), as well as perhaps a callback of `PUT thirdpartyRequests/transactions/{id}/error` from the PayerDFSP, as highlighted with the error conditions `1.1`, `1.2`, and `1.5` above.

## Proposal

In light of the above, I think we should _not_ rely on the switch to generate the `PUT thirdpartyRequests/transactions/{id}` at the end of a 3rd Party Transaction, and instead require it from the PayerDFSP. After all, the PayerDFSP is the one who knows all of the context about the 3rd Party Transaction, and the associated Quote, Authorizations and Request

**Comment by mjbrichards** [comment]
Hi Lewis,
thanks again for your detailed thinking. I'd like to start by looking at your list of failure conditions, and mention the following points:

- 1 and 2 are direct conversations between the PISP and the Payer DFSP. The PISP will be notified of a failure in any case.
- On 4, are we talking about the quote response from the Payee DFSP?
- 5 is an especially interesting one, since it is not generated in response to a message, but I guess occurs when the Payer DFSP finds that it cannot reserve the payer's funds because their account has insufficient funds (although, presumably, it had sufficient funds when the original request was received from the PISP.) So there's no obvious current model for how that message would get generated, or what it would be.

Our general model in Mojaloop is that participants are responsible for (cobbling together) their own understanding of the state of a transaction: for instance, if the payer decides not to proceed with a quotation, it is not under any obligation to inform the payee DFSP of that fact. Similarly, if the payee DFSP expires a quote, it doesn't have to bother sending a ship-has-sailed message to the payer DFSP. Similarly, the PISP will have to maintain its own understanding of the state of a transfer - for instance, connecting a transfer request with the quote it was based on. So we don't have a model in which there is a single transfer document which gets enriched over the course of the process: individual participants exchange messages which are more like RPC requests than RESTful exchanges.

We use PUT responses generally as a one-time response to a specific request, rather than in the standard HTTP way as an update to a server document which different clients hold copies of. This might not be conceptually elegant, but it's the general model for the API at present. So, in the normal run of things, a participant in a transaction will get specific error responses to a particular message: there doesn't need to be a general picture of the transaction as a whole.

I think that requiring the Payer DFSP to forward all responses to the PISP is asking it to do more thna it really needs to. I'm also, I have to confess, attracted to the switch-based version because I think it's a simple general pattern that we could also use in all sorts of other contexts (for instance, cross-network and cross-currency.)

**Comment by millerabel** [comment]
Hi both,

"So we don't have a model in which there is a single transfer document which gets enriched over the course of the process: individual participants exchange messages which are more like RPC requests than RESTful exchanges.”

This is a fundamental design choice for Mojaloop. Each DFSP operates on its own ledger and manages its own transaction state. This is a departure from stateless message switches that provide no assurance to participants, but simply mediate the message exchanges.

In Mojaloop, the hub manages one central source of truth about each transfer as it moves through the process of preparation, fulfillment, and commitment. How the payer arrives at a decision to proceed from quotation to present a prepared payment is not the business of the hub.

Likewise whether a payee agrees to progress a prepared payment by presenting fulfillment is also out of scope.

However, once both the payment condition (from the prepared payment) and the fulfillment are acknowledged by the hub, and presuming they are safe within a practical margin of the expiration limit, the hub will treat fulfillment as permission to clear and the transfer will proceed to commitment (assuming there is no formative error or duplication of the presented messages).

The truth remains always at the center and concerned with transfers, not transaction details. Record keeping by DFSPs remains their own responsibility.

So the passing around of transaction quotations and sharing prepared payments between DFSPs and third parties feels eccentric to the core function of the hub. Passing through the quotation request and response (and having a look at them if scheme rules require enforcement) is a convenience activity to simplify communication topology.

But we would not want to revert to the pure message passing architecture used e.g. in ISO20022 or SWIFT message-based switching systems.

— Miller

On Jun 23, 2020, at 2:36 AM, mjbrichards <notifications@github.com> wrote:


Hi Lewis,
thanks again for your detailed thinking. I'd like to start by looking at your list of failure conditions, and mention the following points:

1 and 2 are direct conversations between the PISP and the Payer DFSP. The PISP will be notified of a failure in any case.
On 4, are we talking about the quote response from the Payee DFSP?
5 is an especially interesting one, since it is not generated in response to a message, but I guess occurs when the Payer DFSP finds that it cannot reserve the payer's funds because their account has insufficient funds (although, presumably, it had sufficient funds when the original request was received from the PISP.) So there's no obvious current model for how that message would get generated, or what it would be.
Our general model in Mojaloop is that participants are responsible for (cobbling together) their own understanding of the state of a transaction: for instance, if the payer decides not to proceed with a quotation, it is not under any obligation to inform the payee DFSP of that fact. Similarly, if the payee DFSP expires a quote, it doesn't have to bother sending a ship-has-sailed message to the payer DFSP. Similarly, the PISP will have to maintain its own understanding of the state of a transfer - for instance, connecting a transfer request with the quote it was based on. So we don't have a model in which there is a single transfer document which gets enriched over the course of the process: individual participants exchange messages which are more like RPC requests than RESTful exchanges.

We use PUT responses generally as a one-time response to a specific request, rather than in the standard HTTP way as an update to a server document which different clients hold copies of. This might not be conceptually elegant, but it's the general model for the API at present. So, in the normal run of things, a participant in a transaction will get specific error responses to a particular message: there doesn't need to be a general picture of the transaction as a whole.

I think that requiring the Payer DFSP to forward all responses to the PISP is asking it to do more thna it really needs to. I'm also, I have to confess, attracted to the switch-based version because I think it's a simple general pattern that we could also use in all sorts of other contexts (for instance, cross-network and cross-currency.)

**Comment by lewisdaly** [comment]
Thanks @millerabel and @MichaelJBRichards for your input. I certainly underestimated the role of the switch here in terms of being an _authority on the state of a transaction_, and that helps me see why the final notification to the PISP must come from the switch, and _not_ the Payer DFSP.

We spoke about this in our PISP design meeting today.

In general, we discussed the following principles:
1. The PISP should not be a _dumb client_ of the DFSP
2. The Switch is the authority on the status of transactions. Therefore, we should prefer to use that authority instead of relying on the PayerDFSP to inform the PISP of the transaction outcome.


And made the following decisions:
1. We double checked the logic and design decisions behind making the final transaction status be sent from the switch
2. We updated this to be simply a `PUT /transfers/{id}` call, which will originate from the Payee DFSP, (or the switch in the case where the Payee DFSP uses the `PATCH /transfers/{id}` call)
3. We decided that, based on our assumptions around the transaction success cases, the switch should also send `PUT /transfers/{id}/error` calls to the PISP based on the transaction subscriptions. This indeed won't cover all failure conditions in the PISP case, so we will also allow for the Payer DFSP to send `PUT /thirdpartyRequests/transactions/{id}/error` callbacks to cover errors that only the DFSP would know about.
4. Regarding the subscription pattern for `PUT /quotes/{id}` callbacks, we will add such a revision to the backlog, and move forward with the current design of using `POST /authorizations` to get the QuoteResponse object to the PISP
